### ISSUE TITLE
ADD brignet

### ISSUE THREAD

**Issue by kyakuno** [axinc-ai/ailia-models#309]
https://github.com/pKrime/brignet
GPL3

**Comment by kyakuno** [comment]
4つのモデルの組み合わせ。

```
    jointNet = JOINTNET()
    jointNet.to(device)
    jointNet.eval()
    jointNet_checkpoint = torch.load(os.path.join(model_dir, 'gcn_meanshift/model_best.pth.tar'))
    jointNet.load_state_dict(jointNet_checkpoint['state_dict'])
    print("     joint prediction network loaded.")

    rootNet = ROOTNET()
    rootNet.to(device)
    rootNet.eval()
    rootNet_checkpoint = torch.load(os.path.join(model_dir, 'rootnet/model_best.pth.tar'))
    rootNet.load_state_dict(rootNet_checkpoint['state_dict'])
    print("     root prediction network loaded.")

    boneNet = BONENET()
    boneNet.to(device)
    boneNet.eval()
    boneNet_checkpoint = torch.load(os.path.join(model_dir, 'bonenet/model_best.pth.tar'))
    boneNet.load_state_dict(boneNet_checkpoint['state_dict'])
    print("     connection prediction network loaded.")

    skinNet = SKINNET(nearest_bone=5, use_Dg=True, use_Lf=True)
    skinNet_checkpoint = torch.load(os.path.join(model_dir, 'skinnet/model_best.pth.tar'))
    skinNet.load_state_dict(skinNet_checkpoint['state_dict'])
    skinNet.to(device)
    skinNet.eval()
    print("     skinning prediction network loaded.")
```

**Comment by kyakuno** [comment]
オリジナルリポジトリ
https://github.com/zhan-xu/RigNet

**Comment by kyakuno** [comment]
bRigNet - 注目のディープラーニングを用いた自動リギング技術！Blender向けにアドオン化した人が現れました！
https://3dnchu.com/archives/brignet/

**Comment by kyakuno** [comment]
(1) onnxへのエクスポート
(2) ailia-models用のサンプルを作成
(3) brignet pluginをailia SDKを使用するように書き換え

**Comment by ooe1123** [comment]
https://github.com/zhan-xu/RigNet の環境セットアップではtorch1.6.0を使っているが、TorchScriptのエクスポートで落ちるため、torch1.7.0を入れる。torch1.7.0では無事にエクスポートできた。
```
pip install numpy scipy matplotlib tensorboard open3d==0.9.0 opencv-python
pip install "rtree>=0.8,<0.9" 
pip install trimesh[easy]
pip install torch==1.7.1+cpu torchvision==0.8.2+cpu torchaudio===0.7.2 -f https://download.pytorch.org/whl/torch_stable.html
pip install torch-scatter -f https://pytorch-geometric.com/whl/torch-1.7.0+cpu.html
pip install torch-sparse -f https://pytorch-geometric.com/whl/torch-1.7.0+cpu.html
pip install torch-cluster -f https://pytorch-geometric.com/whl/torch-1.7.0+cpu.html
pip install torch-spline-conv -f https://pytorch-geometric.com/whl/torch-1.7.0+cpu.html
pip install torch-geometric
```

**Comment by ooe1123** [comment]
gcn_meanshiftモデルをエクスポートするために、以下の対応を行った。
① torch_scatter/scatter.pyのscatter_max()を置き換え
```
@torch.jit.script
def scatter_max(
        src: torch.Tensor, index: torch.Tensor, dim: int = -1,
        out: Optional[torch.Tensor] = None,
        dim_size: Optional[int] = None) -> Tuple[torch.Tensor, torch.Tensor]:

    if dim_size is None:
        dim_size = int(index.max())+1
    min = torch.full(src.size(), -1e6)
    w = src.size(1)
    a = []
    b = []
    for i in range(dim_size):
        x = torch.where(
            (index==i).view(-1,1).repeat(1, w), src, min
            ).max(dim=0)
        a += [ x[0].view(1,-1) ]
        b += [ x[1].view(1,-1) ]
    return (torch.cat(a, 0), torch.cat(b, 0))
```
② RigNet/models/GCN.pyで呼び出しているtorch.bincount()とtorch.repeat_interleave()をそれぞれ以下への呼び出しに変更
```
@torch.jit.script
def torch_bincount(src: torch.Tensor) -> torch.Tensor:
    dim_size = int(src.max()) + 1
    a = []
    ones = torch.ones(src.shape).long()
    zeros = ones * 0
    for i in range(dim_size):
        a += [torch.ones(1).long() * torch.where(src == i, ones, zeros).sum()]
    return torch.cat(a)
```
`x_global = torch.repeat_interleave(x_global, torch.bincount(batch), dim=0)`
↓
`x_global = x_global.repeat(torch_bincount(batch)[0], 1)`

③ torch_geometric/nn/conv/message_passing.pyのMessagePassing.aggregate()の呼び出しで、dim_sizeパラメータがpython変数であるために、Tensorのshapeに合わせた推論ができていない。内部のscatter_max()呼び出しでdim_sizeを取得するように、パラメータをNoneに設定する。
```
aggr_kwargs['dim_size'] = None # <------------- 追記
out = self.aggregate(out, **aggr_kwargs)
```

**Comment by ooe1123** [comment]
scatter_max()の最適化バージョン
```
@torch.jit.script
def torch_scatter_max(
        src: torch.Tensor, index: torch.Tensor,
        dim_size: Optional[int] = None
        ) -> Tuple[torch.Tensor, torch.Tensor]:
    if dim_size is None:
        dim_size = int(index.max())+1
    min = torch.full(src.size(), -1e6)
    n = src.size(0)
    w = src.size(1)
    arange = torch.arange(n)
    src = torch.cat([src, torch.full((1, w), -1e6)], 0)
    a = []
    b = []
    for i in range(dim_size):
        msk = torch.masked_select(arange, index==i)
        msk = torch.cat([msk, torch.full([1], n, dtype=torch.long)])
        x = src[msk].max(dim=0)
        a += [ x[0].view(1,-1) ]
        b += [ x[1].view(1,-1) ]
    return (torch.cat(a, 0), torch.cat(b, 0))
```

**Comment by ooe1123** [comment]
ROOT_GCN.pyとPairCls_GCN.pyで定義されているSAModule、FPModuleが
torch_geometric.nnのfps, radius, knn_interpolate を使っていてONNXにエクスポートできないため、
この処理部分をonnxの外出しにして、モデルを分割してエクスポートするようにする。
